File: lib/tdnoisedifffit.m
function [P,fval,Diagnostic] = tdnoisedifffit(x,varargin)
%TDNOISEDIFFFIT computes MLE parameters for the time-domain noise model
%
% Syntax:   P = tdnoisefit(x,Options)
%
% Description:
%
% TDNOISEFIT computes the noise parameters sigma and the underlying signal
% vector mu for the data matrix x, where the columns of x are each noisy
% measurements of mu.
%  
% Inputs:
%   x               Data matrix
%
% Optional inputs:
%   Options         Fit options
%
% Option fields:
%   v0              Initial guess, noise model parameters [3x1 double]
%   mu0             Initial guess, signal vector [Nx1 double]
%   A0              Initial guess, amplitude vector [Mx1 double]
%   eta0            Initial guess, delay vector [Mx1 double]
%   ts              Sampling time [scalar double]
%   Fix             Fixed variables [struct]
%   Ignore          Ignore variables [struct]
%
% Outputs:
%   P               Output parameter structure
%       .logv       Log of noise parameters
%       .mu         Signal vector
%       .A          Amplitude vector
%       .eta        Delay vector
%       .ts         Samling time
%   fval            Value of NLL cost function from FMINUNC
%   Diagnostic      Structure of diagnostic information
%       .exitflag	Exit flag from FMINUNC
%       .output     Output from FMINUNC
%       .grad     	NLL cost function gradient from FMINUNC
%       .hessian   	NLL cost function hessian from FMINUNC
%

% Parse inputs
[N,M] = size(x);
Input = inputParser;
addRequired(Input,'x', ...
    @(x) validateattributes(x,{'double'},{'2d'}));
addOptional(Input,'Options', struct, ...
    @(Options) validateattributes(Options,{'struct'},{'nonempty'}));
parse(Input,x,varargin{:})
Options = Input.Results.Options;

% Parse options
InputOptions = inputParser;
addParameter(InputOptions,'v0',mean(var(x,1,2))*ones(3,1), ...
    @(v0) validateattributes(v0,{'double'},{'vector','>',0}));
addParameter(InputOptions,'mu0',mean(x,2), ...
    @(mu0) validateattributes(mu0,{'double'},{'vector'}));
addParameter(InputOptions,'A0',ones(M,1), ...
    @(A0) validateattributes(A0,{'double'},{'vector'}));
addParameter(InputOptions,'eta0',zeros(M,1), ...
    @(eta0) validateattributes(eta0,{'double'},{'vector'}));
addParameter(InputOptions,'ts',1, ...
    @(ts) validateattributes(ts,{'numeric'},{'scalar','positive'}));
addParameter(InputOptions,'Fix', ...
    struct('logv',false,'mu',false,'A',true,'eta',true),...
    @(Fix) validateattributes(Fix,{'struct'},{'nonempty'}))
addParameter(InputOptions,'Ignore', ...
    struct('A',true,'eta',true),...
    @(Ignore) validateattributes(Ignore,{'struct'},{'nonempty'}))
parse(InputOptions,Options)

v0 = InputOptions.Results.v0(:);
mu0 = InputOptions.Results.mu0(:);
A0 = InputOptions.Results.A0(:);
eta0 = InputOptions.Results.eta0(:);
ts = InputOptions.Results.ts;
Fix = InputOptions.Results.Fix;
Ignore = InputOptions.Results.Ignore;

% Define optimization problem
MLE.solver = 'fminunc';
MLE.options = optimoptions('fminunc',...
    'SpecifyObjectiveGradient',false,...
    'Algorithm','quasi-newton',...
    'FiniteDifferenceType','central',...
    'FiniteDifferenceStepSize', 1e-7,...
    'StepTolerance',1e-8,...
    'OptimalityTolerance',1e-8,...
    'MaxFunctionEvaluations',1e6,...
    'Display','off',...
    'UseParallel',true,...
    'Diagnostics','off');
% ,...
%     'FiniteDifferenceType','central',...
%     'FiniteDifferenceStepSize', 1e-8,...
%     'SubproblemAlgorithm', 'cg', ...
%     'FiniteDifferenceType','central',...
%     'FiniteDifferenceStepSize', 1e-7,...
%     'CheckGradients',true,...
%     'StepTolerance',1e-8,...
%     'FunctionTolerance',1e-8,...
%     'OptimalityTolerance',1e-8,...
%     'MaxFunctionEvaluations',1e6,...
%     'UseParallel',true,...
MLE.x0 = [];
idxStart = 1;
if Fix.logv
    setPlogv = @(p) log(v0);
else
    MLE.x0 = [MLE.x0;log(v0)];
    idxEnd = idxStart+2;
    idxRange = idxStart:idxEnd;
    setPlogv = @(p) p(idxRange);
    idxStart = idxEnd + 1;
end
if Fix.mu
    setPmu = @(p) mu0;
else
    MLE.x0 = [MLE.x0;mu0];
    idxEnd = idxStart+N-1;
    idxRange = idxStart:idxEnd;
    setPmu = @(p) p(idxRange);
    idxStart = idxEnd + 1;
end
if Ignore.A
    setPA = @(p) [];
elseif Fix.A
    setPA = @(p) A0(:);
else
    MLE.x0 = [MLE.x0;A0(2:end)];
    idxEnd = idxStart+M-2;
    idxRange = idxStart:idxEnd;
    setPA = @(p) [A0(1);p(idxRange)];
    idxStart = idxEnd + 1;
end
if Ignore.eta
    setPeta = @(p) [];
elseif Fix.eta
    setPeta = @(p) eta0(:);
else
    MLE.x0 = [MLE.x0;eta0(2:end)];
    idxEnd = idxStart+M-2;
    idxRange = idxStart:idxEnd;
    setPeta = @(p) [eta0(1);p(idxRange)];
end

D = tdtf(@(theta,w) -1i*w,0,N,ts);

parseIn = @(p) struct('logv',setPlogv(p),'mu',setPmu(p),...
    'A',setPA(p),'eta',setPeta(p),'ts',ts,'D',D);
MLE.objective = @(p) tdnllfixedeffects(x,parseIn(p),Fix);

[pOut,fval,exitflag,output,grad,hessian] = fminunc(MLE);

% Parse output
idxStart = 1;
if Fix.logv
    P.var = v0;
else
    idxEnd = idxStart+2;
    idxRange = idxStart:idxEnd;
    idxStart = idxEnd + 1;
    P.var = exp(pOut(idxRange));
end
if Fix.mu
    P.mu = mu0;
else
    idxEnd = idxStart+N-1;
    idxRange = idxStart:idxEnd;
    idxStart = idxEnd + 1;
    P.mu = pOut(idxRange);
end
if Ignore.A || Fix.A
    P.A = A0;
else
    idxEnd = idxStart+M-2;
    idxRange = idxStart:idxEnd;
    idxStart = idxEnd + 1;
    P.A = [A0(1);pOut(idxRange)];
end
if Ignore.eta || Fix.eta
    P.eta = eta0;
else
    idxEnd = idxStart+M-2;
    idxRange = idxStart:idxEnd;
    P.eta = [eta0(1);pOut(idxRange)];
end

P.ts = ts;

if nargout > 2
    Diagnostic.exitflag = exitflag;
    Diagnostic.output = output;
    Diagnostic.grad = grad;
    Diagnostic.hessian = hessian;
    Diagnostic.Err = struct('var',[],'mu',[],'A',[],'eta',[]);
    
    paramTest = ~[Fix.logv;...
        Fix.mu;...
        (Fix.A || Ignore.A) ;...
        (Fix.eta || Ignore.eta)];
    
    Nparam = sum(paramTest.*[3;N;M-1;M-1]);
    V = speye(Nparam)/Diagnostic.hessian;
    err = sqrt(diag(V));
    
    idxStart = 1;
    if paramTest(1)
        Diagnostic.Err.var = ...
            sqrt(diag(diag(P.var)*V(1:3,1:3)*diag(P.var)));
        idxStart = idxStart + 3;
    end
    if paramTest(2)
        Diagnostic.Err.mu = err(idxStart+(0:N-1));
        idxStart = idxStart + N;
    end
    if paramTest(3)
        Diagnostic.Err.A = err(idxStart+(0:M-2));
        idxStart = idxStart + M - 1;
    end
    if paramTest(4)
        Diagnostic.Err.eta = err(idxStart+(0:M-2));
    end

end

%% TDNLLFIXEDEFFECTS
    function [nll,gradnll] = tdnllfixedeffects(x,P,Fix)
    % Strip first element of gradient vector wrt A and eta
        
        [nll,gradnll] = tdnll(x,P,Fix);
        idxStart = 1;
        if ~Fix.logv
            idxStart = idxStart + 3;
        end
        if ~Fix.mu
            idxStart = idxStart + N;
        end
        if ~Fix.A
            gradnll(idxStart) = [];
            idxStart = idxStart + M - 1;
        end
        if ~Fix.eta
            gradnll(idxStart) = [];
            idxStart = idxStart + M - 1;
        end        
    end

end

File: thztoolsPY/tdnoisedifffit.py
import numpy as np
from scipy.optimize import minimize

from thztoolsPY.tdnll import tdnll
from thztoolsPY.tdtf import tdtf


def tdnoisefffit(x, param, fix= {'logv': False, 'mu': False, 'a': True, 'eta': True}, ignore={'a': True, 'eta': True}):
    """ Computes maximum likelihood parameters for the time-domain noise model.

     Tdnosefffit computes the noise parameters sigma and the underlying signal
     vector mu for the data matrix x, where the columns of x are each noisy
     measurements of mu.

    Parameters
    ----------

    x : ndarray or matrix
        Data matrix.

    param : dict
        A dictionary containing initial guess for the parameters:

        v0 : ndarray
            Initial guess, noise model parameters. Array of real elements of size (3, )

        mu0 : ndarray
            Initial guess, signal vector. Array of real elements of size  (n, )

        a0 : ndarray
            Initial guess, amplitude vector. Array of real elements of size (m, )

        eta0 : ndarray
            Initial guess, delay vector. Array of real elements of size (m, )

        ts : int
            Sampling time

    fix : dict, optional
        Fixed variables. If not given, chosen to set free all the variables.

    ignore : dict, optional
        Ignore variables. If not given, chosen to ignore both amplitude and delay.

    Returns
    --------
    p : dict
        Output parameter dictionary containing:
            eta : ndarray
                Delay vector.

            a : ndarray
                Amplitude vector.

            mu : ndarray
                Signal vector.

            var : ndarray
                Log of noise parameters


        fval : float
           Value of NLL cost function from FMINUNC

        Diagnostic : dict
            Dictionary containing diagnostic information
                err : dic
                    Dictionary containing  error of the parameters.

                grad : ndarray
                      Negative loglikelihood cost function gradient from scipy.optimize.minimize BFGS method.

                hessian : ndarray
                    Negative loglikelihood cost function hessian from scipy.optimize.minimize BFGS method.

      """

    n, m = x.shape
    mle = {}
    idxstart = 0

    if fix['logv']:

        def setplogv(p):
            return np.log(param['v0'])
    else:
        mle['x0'] = [np.log(param['v0'])]
        idxend = idxstart + 2
        idxrange = np.arange(idxstart, idxend + 1)

        def setplogv(p):
            return p[idxrange]

        idxstart = idxend + 1
    pass

    if fix['mu']:
        def setpmu(p):
            return param['mu0']
    else:
        mle['x0'] = [param['mu0']]
        idxend = idxstart + n - 1
        idxrange = np.arange(idxstart, idxend + 1)

        def setpmu(p):
            return p[idxrange]

        idxstart = idxend + 1
    pass

    if ignore['a']:
        def setpa(p):
            return []

    elif fix['a']:
        def setpa(p):
            return param['a0']

    else:
        mle['x0'] = param['a0'][1:]
        idxend = idxstart + m - 2
        idxrange = np.arange(idxstart, idxend + 1)

        def setpa(p):
            return p[idxrange]

        idxstart = idxend + 1
    pass

    if ignore['eta']:
        def setpeta(p):
            return []

    elif fix['eta']:
        def setpeta(p):
            return param['eta']

    else:
        mle['x0'] = [param['eta0']][1:]
        idxend = idxstart + m - 1
        idxrange = np.arange(idxstart, idxend + 1)

        def setpeta(p):
            return np.concatenate((np.array([param['eta']][0]), p(idxrange)))
    pass

    def fun(theta, w):
        return -1j * w

    d = tdtf(fun, 0, n, param['ts'])

    def parsein(p):
        return {'logv': setplogv(p), 'mu': setpmu(p), 'a': setpa(p), 'eta': setpeta(p), 'ts': param['ts'], 'd': d}

    def tdnllfixedeffects(x, param, fix):
        """ Strip first element of gradient vector wrt A and eta """

        nll, gradnll = tdnll(x, param, fix)
        idxstart = 0
        if np.logical_not(fix['logv']):
            idxstart = idxstart + 3
        pass

        if np.logical_not(fix['mu']):
            idxstart = idxstart + n
        pass

        if np.logical_not(fix['a']):
            gradnll[idxstart] = []
            idxstart = idxstart + m - 1
        pass

        if np.logical_not(fix['eta']):
            gradnll[idxstart] = []
        pass

        return nll, gradnll

    def objective(p):
        return tdnllfixedeffects(x, parsein(p), fix)

    mle['objective'] = objective

    out = minimize(mle['objective'], mle['x0'], method='BFGS')

    # Parse output
    p = {}

    idxstart = 0
    if fix['logv']:
        p['var'] = param['v0']

    else:
        idxend = idxstart + 3
        idxrange = np.arange(idxstart, idxend + 1)
        p['var'] = np.exp(out.x[idxrange])
    pass

    if fix['mu']:
        p['mu'] = param['mu0']

    else:
        idxend = idxstart + n - 1
        idxrange = np.arange(idxstart, idxend + 1)
        idxstart = idxend + 1
        p['mu'] = out.x[idxrange]
    pass

    if ignore['a'] or fix['a']:
        p['a'] = param['a0']

    else:
        idxend = idxstart + m - 2
        idxrange = np.arange(idxstart, idxend + 1)
        idxstart = idxend + 1
        p['a'] = np.concatenate((np.array([param['a0']][0]), out.x(idxrange)))
    pass

    if ignore['eta'] or fix['eta']:
        p['eta'] = param['eta0']

    else:
        idxend = idxstart + m - 2
        idxrange = np.array(idxstart, idxend + 1)
        p['eta'] = np.concatenate((np.array([param['eta']][0]), out.x(idxrange)))
    pass

    p['ts'] = param['ts']

    diagnostic = {'grad': out.jac, 'hessian': np.linalg.inv(out.hess_inv),
                  'err': {'var': [], 'mu': [], 'a': [], 'eta': []}}

    paramtest = np.logical_not([fix['logv'], fix['mu'], fix['a'] or ignore['a'], fix['eta'] or ignore['eta']])

    ngrad = np.sum(paramtest * [[3], [n], [m - 1], [m - 1]])
    v = np.identity(ngrad) / diagnostic['hessian']
    err = np.sqrt(np.diag(v))

    idxstart = 0
    if paramtest[0]:
        diagnostic['err']['var'] = np.sqrt(np.diag(np.diag(p['var']) * v[0:3, 0:3]) * np.diag(p['var']))
        idxstart = idxstart + 3
    pass

    if paramtest[1]:
        diagnostic['err']['mu'] = err[idxstart, (idxstart + n - 1)]
        idxstart = idxstart + n
    pass

    if paramtest[2]:
        diagnostic['err']['a'] = err[idxstart, (idxstart + m - 2)]
        idxstart = idxstart + m - 1
    pass

    if paramtest[3]:
        diagnostic['err']['eta'] = err(idxstart, (idxstart + m - 2))
    pass

    return p, out.fun, diagnostic
